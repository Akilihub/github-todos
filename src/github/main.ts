/**
 * This is the entry point for your Probot App.
 * @param {import('probot').Application} app - Probot's Application class.
 */
export const main = (app: any) => {
    app.log('Cheers, the app runs on a server!')
  
    app.on('push', async (context: any) => {
      const postIssue: string = context.issue({body: 'Thanks for opening a repository in this account.'})
      return context.github.issues.createComment(postIssue)
    })
  
  }
  