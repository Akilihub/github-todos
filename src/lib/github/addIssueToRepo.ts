/**
 *  API Function that adds an issue with comments to a repository 
 */

 export interface IssueWithComments {
    title: string,
    comments: string[]
 }

 export default async function addIssueToRepo(context: any): Promise<IssueWithComments[]> {
    const octokit = context.github;

    const fields = {
      owner: "author",
      repo: "repository",
      title: "title",
      body: "comments"
    };

    const result = await octokit.issues.create(fields); // ** ??? get comments
    return result.issues.map( issue => {
      return {
        title: issue.title,
        comments: issue.comments
      }
    })
 }