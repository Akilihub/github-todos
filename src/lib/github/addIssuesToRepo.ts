/**
 * API function to add comments to existing bot created issues.
 */
import { find } from "ramda";
import { getBasicRepoProps } from "./utils";

import { Context, Issue } from "./types";
import { getAllRepoIssues } from "./repoIssuesList";

export default async function addIssuesToRepo (context: Context, newIssues: Issue[]): Promise<void> {
  const octokit = context.github;
  const { owner, repo } = getBasicRepoProps (context);
  const ghIssues = await getAllRepoIssues(context);
  newIssues.forEach(async ({title, body}) => {
    const currentGHIssue = find(ghIssue => ghIssue.title === title, ghIssues);
    const fields = { // TODO: maybe adde assignees
      owner,
      repo,
      body,
      title,
      labels: ["GH-TODO-BOT"],
      ...(currentGHIssue ? { number: currentGHIssue.number } : {})
    };
    return currentGHIssue
      ? octokit.issues.createComment(fields)
      : octokit.issues.create(fields);
  });

}
