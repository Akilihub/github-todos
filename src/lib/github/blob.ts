import fetch from "node-fetch";

export interface Blob {
    content: string;
    name: string;
}

const getFileContent = async (arrList): Promise<Blob[]> => {
  const promises: Array<Promise<Blob>> = arrList.map( async (file) => {
    const res = await fetch(file.url);
    const textBlob = await res.text();
    return ({
      content: textBlob,
      name: file.name
    });
  });
  return Promise.all(promises);
};

export default getFileContent;
