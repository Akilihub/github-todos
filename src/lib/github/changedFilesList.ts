/**
 * Gets files that have changed during a commit
 */
import { Context } from "./types";
import { getBasicRepoProps } from "./util";

export interface ModifiedFile {
  name: string;
  url: string;
}

export default async function getChangedFiles(context: Context): Promise<ModifiedFile[]> {
  const repoPropsObj = getBasicRepoProps (context);
  const sha = context.payload.head_commit.id;
  const octokit = context.github;
  const fields = {
    owner: repoPropsObj.owner,
    repo: repoPropsObj.repo,
    sha
  };
  const result = await octokit.repos.getCommit( fields);
  return result.files.map(file => ({name: file.filename, url: file.blob_url}));
}