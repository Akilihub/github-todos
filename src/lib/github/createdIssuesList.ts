/**
 * extracts a list of open issues created by the app
 */

export interface openIssues {
    title: string; // issue title
    id: string; // issue id
};

export default async function autoCreatedIssues (context: any): Promise<openIssues> {
    const octokit = context.github;
    const opts = {
      filter: 'subscribed', 
      state: 'open'
    }
    const result = await octokit.issues.getAll(opts);
    return result.files.map(file => ( {name: file.title, url: file.number} ));
}