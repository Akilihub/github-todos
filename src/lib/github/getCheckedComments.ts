import { getBasicRepoProps } from "./utils";

import { Context } from "./types";

export default async function getCheckedComments (context: Context): Promise<string[]> {
  const octokit = context.github;
  const { owner, repo } = getBasicRepoProps (context);
  const fields = {
      owner,
      repo,
      labels: ["GH-TODO-BOT"],
    };
  const arrayComments = await octokit.octokit.issues.getForRepo(fields);
  return arrayComments.map(obj =>
    obj.body.split(`\n`).filter( (comment: string) => comment.includes(`[x]`)));
}
