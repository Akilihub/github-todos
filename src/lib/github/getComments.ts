import { getBasicRepoProps } from "./utils";

import { Context } from "./types";
import { labels } from "./addIssuesToRepo";

export interface GhIssueComment {
  file: string;
  comment: string;
  assignee: string;
}

export function getIssueCommentDetails (strng: String): GhIssueComment {
  const arr = strng.replace(/\s/g, "")
  .replace(/[\[\]-]+/g, "")
  .replace(/\(.*\)/, "").split(":");
  return {
    file: arr[0],
    comment: arr[1],
    assignee: arr[3]
  };
}

export default async function getAllComments (context: Context): Promise<GhIssueComment[]> {
  const octokit = context.github;
  const { owner, repo } = getBasicRepoProps (context);
  const fields = {
      owner,
      repo,
      labels: labels,
    };
  const arrayComments = await octokit.issues.list(fields);
  return arrayComments.map(obj => {
    const commentArray: string[] = obj.body.split(`\n`);
    return commentArray.map(getIssueCommentDetails);
  });
}

