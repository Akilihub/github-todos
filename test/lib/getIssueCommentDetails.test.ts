import "jest";
import { getIssueCommentDetails } from "../../src/lib/github/getComments";

const strng = ` - [ ] [app.js](https://github.com/aleku399/sequelizeDemoApp/blob/master/app.js#L36) : testing the bot: TODO : aleku399`;

describe("should return an array of file and comment", () => {
    it("should return array", () => {
      const results = getIssueCommentDetails(strng);
      expect(results).toMatchSnapshot();
    });
});

