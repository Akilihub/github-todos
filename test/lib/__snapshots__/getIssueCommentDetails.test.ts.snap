// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return an array of file and comment should return array 1`] = `
Object {
  "assignee": "aleku399",
  "comment": "testingthebot",
  "file": "app.js",
}
`;
